File: PAT_Basic_Level_by_c/1013 数素数/1013 数素数.c
/*
 * 题目：1013 数素数
 * 实现原理：
 *   输出第m个到第n个素数，每行输出10个。
 *   算法步骤：
 *   1. 实现is_prime函数判断素数（优化：只检查奇数因子到sqrt(n)）
 *   2. 从2开始逐个检查每个数是否为素数
 *   3. 记录素数的序号，当序号在[m, n]范围内时输出
 *   4. 使用print_count控制每行输出10个素数，最后一个素数后不输出空格
 */
#include <stdio.h>
#include <math.h>

int is_prime(int n) {
    if (n <= 1) return 0;                   // 0、1 不是素数
    if (n == 2) return 1;                   // 2 是唯一的偶素数
    if (n % 2 == 0) return 0;               // 其他偶数都不是素数
    for (int i = 3; i <= sqrt(n); i += 2) { // 只检查奇数因子，到 sqrt(n) 为止
        if (n % i == 0) return 0;           // 能被整除说明不是素数
    }
    return 1;
}

int main() {
    int m, n;
    scanf("%d %d", &m, &n);
    int count = 0, num = 2;     // count 为已找到素数的序号，num 为当前待判断的数
    int print_count = 0;        // 当前行已输出的素数个数（每行 10 个）
    while (count < n) {         // 一直找，直到找到第 n 个素数
        if (is_prime(num)) {
            count++;            // 素数序号加 1
            if (count >= m) {   // 序号在 [m, n] 区间内的才输出
                printf("%d", num);
                print_count++;
                if (print_count % 10 == 0) {    // 每输出 10 个换行
                    printf("\n");
                } else if (count < n) {         // 不是最后一个素数时用空格分隔
                    printf(" ");
                }
            }
        }
        num++;
    }
    if (print_count % 10 != 0) {    // 末行不足 10 个时补一个换行
        printf("\n");
    }
    return 0;
}

File: PAT_Basic_Level_by_go/1013 数素数/1013 数素数.go
// 题目：1013 数素数
// 实现原理：输出第 M 到第 N 个素数（M、N 为正整数且 M <= N）。
// 从 2 开始遍历自然数，每找到一个素数就计数，当计数达到 M 时开始记录，
// 直到计数达到 N 为止。每 10 个素数一行输出。

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// isPrime 判断 n 是否为素数。
// 返回值：true 表示是素数，false 表示不是素数。
func isPrime(n int) bool {
	if n < 2 {
		return false
	}
	// 只需检查到 sqrt(n) 即可
	for i := 2; i*i <= n; i++ {
		if n%i == 0 {
			return false
		}
	}
	return true
}

// main 是程序入口函数，读取 M 和 N，输出第 M 到第 N 个素数。
func main() {
	scanner := bufio.NewScanner(os.Stdin)
	if !scanner.Scan() {
		return
	}
	f := strings.Fields(scanner.Text())
	m, _ := strconv.Atoi(f[0])
	n, _ := strconv.Atoi(f[1])

	count := 0
	num := 2
	out := make([]string, 0)
	// 遍历自然数，直到找到第 N 个素数
	for count < n {
		if isPrime(num) {
			count++
			// 从第 M 个开始记录
			if count >= m {
				out = append(out, strconv.Itoa(num))
			}
		}
		num++
	}
	// 每 10 个一行输出
	var b strings.Builder
	for i, v := range out {
		if i > 0 {
			if i%10 == 0 {
				b.WriteByte('\n')
			} else {
				b.WriteByte(' ')
			}
		}
		b.WriteString(v)
	}
	fmt.Print(b.String())
}